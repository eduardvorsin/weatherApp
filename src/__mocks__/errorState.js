import fetchingStatuses from '../redux/fetchingStatuses';

const errorData = {
  data: {},
  loadingStatus: fetchingStatuses.error,
  error: 'mock error',
};

const errorState = {
  weather: {
    hourlyWeather: errorData,
    currentWeather: errorData,
    dailyWeather: errorData,
    humidity: errorData,
    wind: errorData,
  },
  geocoding: errorData,
};

export default errorState;
