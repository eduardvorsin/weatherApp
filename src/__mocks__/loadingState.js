import fetchingStatuses from '../redux/fetchingStatuses';

const loadingData = {
  data: {},
  loadingStatus: fetchingStatuses.loading,
  error: null,
};

const loadingState = {
  weather: {
    hourlyWeather: loadingData,
    currentWeather: loadingData,
    dailyWeather: loadingData,
    humidity: loadingData,
    wind: loadingData,
  },
  geocoding: loadingData,
};

export default loadingState;
