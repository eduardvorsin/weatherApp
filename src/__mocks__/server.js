import { rest } from 'msw';
import { setupServer } from 'msw/node';
import mockGeocodingError from './geocodingError';
import mockWeatherError from './weatherError';

const isCurrentWeatherOrWindRequest = (searchParams) => (
  searchParams.get('current_weather') === 'true'
);

const isHourlyWeatherRequest = (searchParams) => {
  const currentParams = searchParams.getAll('hourly');
  return currentParams.includes('temperature_2m') && currentParams.includes('weathercode');
};

const isDailyWeatherRequest = (searchParams) => {
  const currentParams = searchParams.getAll('daily');
  return currentParams.includes('temperature_2m_max') && currentParams.includes('temperature_2m_min') && currentParams.includes('weathercode');
};

const isHumidityRequest = (searchParams) => {
  const currentParams = searchParams.getAll('hourly');
  return currentParams.includes('apparent_temperature') && currentParams.includes('relativehumidity_2m');
};

const currentWeatherData = {
  current_weather: {
    weathercode: 1,
    temperature: -6.7,
    winddirection: 128,
    windspeed: 1.64,
  },
};
const hourlyWeatherData = {
  hourly: {
    time: [
      '2022-12-27T03:00',
      '2022-12-27T04:00',
      '2022-12-27T05:00',
      '2022-12-27T06:00',
      '2022-12-27T07:00',
      '2022-12-27T08:00',
      '2022-12-27T09:00',
      '2022-12-27T10:00',
      '2022-12-27T11:00',
      '2022-12-27T12:00',
      '2022-12-27T13:00',
      '2022-12-27T14:00',
      '2022-12-27T15:00',
      '2022-12-27T16:00',
      '2022-12-27T17:00',
      '2022-12-27T18:00',
      '2022-12-27T19:00',
      '2022-12-27T20:00',
      '2022-12-27T21:00',
      '2022-12-27T22:00',
      '2022-12-27T23:00',
      '2022-12-28T00:00',
      '2022-12-28T01:00',
      '2022-12-28T02:00',
      '2022-12-28T03:00',
      '2022-12-28T04:00',
      '2022-12-28T05:00',
      '2022-12-28T06:00',
      '2022-12-28T07:00',
      '2022-12-28T08:00',
    ],
    temperature_2m: [
      -4.5, -4.4, -4.3, -4.5, -5.1, -5.6, -5.8, -6, -6.5, -6.8, -7, -6.4,
      -5.3, -4.3, -3.6, -3.4, -3.7, -4.4, -4.9, -5.3, -5.6, -5.8, -5.7, -5.5,
      -5.0, -4.5, -3.9, -3.1, -3.4, -4.9,
    ],
    weathercode: [1, 2, 3, 71, 71, 3, 3, 2, 2, 2, 1, 3, 71, 86, 1, 1, 3, 2, 2,
      71, 71, 3, 2, 1, 71, 2, 3, 1, 2, 71],
  },
};
const dailyWeatherData = {
  daily: {
    time: ['2022-12-26', '2022-12-27', '2022-12-29', '2022-12-30', '2022-12-31', '2023-01-01', '2023-01-02'],
    temperature_2m_max: [-9.4, -8, -7.3, -5, -6.7, -7.6, -9.8],
    temperature_2m_min: [-17.1, -15.5, -16.4, -17.5, -15.7, -16.3, -13.6],
    weathercode: [3, 2, 71, 2, 2, 3, 71],
  },
};
const humidityData = {
  hourly: {
    time: [
      '2022-12-27T00:00',
      '2022-12-27T01:00',
      '2022-12-27T02:00',
      '2022-12-27T03:00',
      '2022-12-27T04:00',
      '2022-12-27T05:00',
      '2022-12-27T06:00',
      '2022-12-27T07:00',
      '2022-12-27T08:00',
      '2022-12-27T09:00',
    ],
    apparent_temperature: [-17.4, -18, -18.6, -19.1, -19.5, -19, -17, -16, -15, 14],
    relativehumidity_2m: [81, 82, 83, 84, 85, 86, 71, 72, 73, 74],
  },
};
const geocodingData = {
  address: {
    city: 'Вестминстер',
    state_district: 'Большой Лондон',
    state: 'Англия',
    country: 'Великобритания',
    country_code: 'gb',
    'ISO3166-2-lvl4': 'GB-ENG',
    'ISO3166-2-lvl8': 'GB-WSM',
  },
};

const handlers = [
  rest.get('https://api.open-meteo.com/v1/forecast', (req, res, ctx) => {
    const { searchParams } = req.url;

    if (isCurrentWeatherOrWindRequest(searchParams)) {
      return res(ctx.json(currentWeatherData), ctx.delay(150));
    }
    if (isHourlyWeatherRequest(searchParams)) {
      return res(ctx.json(hourlyWeatherData), ctx.delay(150));
    }
    if (isDailyWeatherRequest(searchParams)) {
      return res(ctx.json(dailyWeatherData), ctx.delay(150));
    }
    if (isHumidityRequest(searchParams)) {
      return res(ctx.json(humidityData), ctx.delay(150));
    }

    return res(ctx.json('unknown request'), ctx.delay(150));
  }),
  rest.get('https://nominatim.openstreetmap.org/reverse', (_, res, ctx) => (
    res(ctx.json(geocodingData), ctx.delay(150))
  )),
];

export const errorHandlers = [
  rest.get('https://nominatim.openstreetmap.org/reverse', (_, res, ctx) => res(ctx.status(404), ctx.json(mockGeocodingError), ctx.delay(150))),
  rest.get('https://api.open-meteo.com/v1/forecast', (_, res, ctx) => res(ctx.status(404), ctx.json(mockWeatherError), ctx.delay(150))),
];

const server = setupServer(...handlers);

export const setupServerWithErrors = () => server.use(...errorHandlers);

export default server;
