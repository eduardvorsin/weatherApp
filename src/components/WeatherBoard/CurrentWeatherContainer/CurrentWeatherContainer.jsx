import PropTypes from 'prop-types';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import fetchingStatuses from '../../../redux/fetchingStatuses';
import selectCurrentWeather from '../../../redux/selectors/selectCurrentWeather/selectCurrentWeather';
import fetchCurrentWeather from '../../../redux/thunks/fetchCurrentWeather/fetchCurrentWeather';
import ErrorInfo from '../../UI/ErrorInfo/ErrorInfo';
import Spinner from '../../UI/Spinner/Spinner';
import CurrentWeather from '../CurrentWeather/CurrentWeather';

export default function CurrentWeatherContainer({
  className,
  location,
  testId,
}) {
  const dispatch = useDispatch();

  const {
    data,
    loadingStatus,
  } = useSelector(selectCurrentWeather);

  useEffect(() => {
    if (location.latitude && location.longitude) {
      dispatch(fetchCurrentWeather(location));
    }
  }, [location, dispatch]);

  return (
    <section
      className={`current-weather__section current-weather-section ${className}`}
      data-testid={testId}
    >
      <h2 className="current-weather-section__title sr-only">
        Текущая погода
      </h2>
      {loadingStatus === fetchingStatuses.loading
        && (
          <Spinner
            size={102}
          />
        )}
      {
        loadingStatus === fetchingStatuses.error
        && (
          <ErrorInfo
            message="Не удалось загрузить данные о текущей погоде"
            headingLevel={3}
          />
        )
      }
      {
        loadingStatus === fetchingStatuses.resolved
        && (
          <CurrentWeather
            weatherCode={data.weatherCode}
            temperature={data.temperature}
          />
        )
      }
    </section>
  );
}

CurrentWeatherContainer.defaultProps = {
  className: '',
  location: {},
  testId: '',
};

CurrentWeatherContainer.propTypes = {
  className: PropTypes.string,
  testId: PropTypes.string,
  location: PropTypes.shape({
    latitude: PropTypes.number,
    longitude: PropTypes.number,
  }),
};
