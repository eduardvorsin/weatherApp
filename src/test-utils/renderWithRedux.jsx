import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createReduxStore } from '../redux/store';

export default function renderWithRedux(component, initialState) {
  const store = createReduxStore(initialState);

  return render(
    <Provider store={store}>
      {component}
    </Provider>,
  );
}
