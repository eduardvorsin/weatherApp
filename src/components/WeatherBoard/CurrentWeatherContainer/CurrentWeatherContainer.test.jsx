import { screen, waitForElementToBeRemoved } from '@testing-library/react';
import renderWithRedux from '../../../test-utils/renderWithRedux';
import errorState from '../../../__mocks__/errorState';
import loadingState from '../../../__mocks__/loadingState';
import mockLocation from '../../../__mocks__/location';
import resolvedState from '../../../__mocks__/resolvedState';
import server, { setupServerWithErrors } from '../../../__mocks__/server';
import CurrentWeatherContainer from './CurrentWeatherContainer';

const temperatureRegexp = /(-)?\d+(.\d)?°/;
const weatherStatusRegexp = /ясно|пасмурно|туман|дождь|снег|гроза/;

describe('CurrentWeatherContainer component tests', () => {
  it('renders correctly', () => {
    renderWithRedux(<CurrentWeatherContainer
      className="test-class"
      location={{}}
      testId="current-weather"
    />);

    expect(screen.getByTestId('current-weather')).toBeInTheDocument();
  });

  it('renders with error when loadingStatus equals "error"', () => {
    renderWithRedux(
      <CurrentWeatherContainer
        className="test-class"
        location={{}}
      />,
      errorState,
    );

    expect(screen.getByText(/Не удалось загрузить данные о текущей погоде/i)).toBeInTheDocument();
    expect(screen.queryByText(weatherStatusRegexp)).not.toBeInTheDocument();
    expect(screen.queryByText(temperatureRegexp)).not.toBeInTheDocument();
  });

  it('renders with loading spinner when loadingStatus equals "loading"', () => {
    renderWithRedux(
      <CurrentWeatherContainer
        className="test-class"
        location={{}}
      />,
      loadingState,
    );

    expect(screen.getByTitle('loading spinner')).toBeInTheDocument();
  });

  it('renders with data when loadingStatus equals "resolved"', () => {
    renderWithRedux(
      <CurrentWeatherContainer
        className="test-class"
        location={{}}
      />,
      resolvedState,
    );

    expect(screen.getByText(weatherStatusRegexp)).toBeInTheDocument();
    expect(screen.getByText(temperatureRegexp)).toBeInTheDocument();
  });

  it('the classname is correctly assigned to the element', () => {
    renderWithRedux(
      <CurrentWeatherContainer
        className="test-class"
        location={{}}
        testId="current-weather"
      />,
    );

    expect(screen.getByTestId('current-weather')).toHaveClass('test-class');
  });

  it('snapshot without anything', () => {
    renderWithRedux(
      <CurrentWeatherContainer
        className="test-class"
        location={{}}
        testId="current-weather"
      />,
    );
    expect(screen.getByTestId('current-weather')).toMatchSnapshot();
  });

  it('snapshot with mock error', () => {
    renderWithRedux(
      <CurrentWeatherContainer
        className="test-class"
        location={{}}
        testId="current-weather"
      />,
      errorState,
    );

    expect(screen.getByTestId('current-weather')).toMatchSnapshot();
  });

  it('snapshot with mock data', () => {
    renderWithRedux(
      <CurrentWeatherContainer
        className="test-class"
        location={{}}
        testId="current-weather"
      />,
      resolvedState,
    );

    expect(screen.getByTestId('current-weather')).toMatchSnapshot();
  });

  it('snapshot with a spinner', () => {
    renderWithRedux(
      <CurrentWeatherContainer
        className="test-class"
        location={{}}
        testId="current-weather"
      />,
      loadingState,
    );

    expect(screen.getByTestId('current-weather')).toMatchSnapshot();
  });
});

describe('CurrentWeatherContainer component itegration tests', () => {
  beforeAll(() => {
    jest.useFakeTimers({
      doNotFake: [
        'hrtime',
        'nextTick',
        'performance',
        'queueMicrotask',
        'requestAnimationFrame',
        'cancelAnimationFrame',
        'requestIdleCallback',
        'cancelIdleCallback',
        'setImmediate',
        'clearImmediate',
        'setInterval',
        'clearInterval',
        'setTimeout',
        'clearTimeout',
      ],
    });
    jest.setSystemTime(new Date('2022-12-27T09:00:00'));

    server.listen({ onUnhandledRequest: 'error' });
  });
  afterEach(() => server.resetHandlers());
  afterAll(() => {
    jest.useRealTimers();

    server.close();
  });

  it('the loading status changes to the error status', async () => {
    setupServerWithErrors();

    renderWithRedux(
      <CurrentWeatherContainer
        location={mockLocation}
      />,
    );

    expect(screen.getByTitle('loading spinner')).toBeInTheDocument();
    expect(screen.queryByText(weatherStatusRegexp)).not.toBeInTheDocument();
    expect(screen.queryByText(temperatureRegexp)).not.toBeInTheDocument();

    await waitForElementToBeRemoved(screen.getAllByTitle('loading spinner'));

    expect(screen.queryByTitle('loading spinner')).not.toBeInTheDocument();
    expect(screen.queryByText(weatherStatusRegexp)).not.toBeInTheDocument();
    expect(screen.queryByText(temperatureRegexp)).not.toBeInTheDocument();
    expect(screen.queryByText(/не удалось загрузить данные о текущей погоде/i)).toBeInTheDocument();
  });

  it('the loading status changes to the resolved status', async () => {
    renderWithRedux(
      <CurrentWeatherContainer
        location={mockLocation}
      />,
    );

    expect(screen.getByTitle('loading spinner')).toBeInTheDocument();
    expect(screen.queryByText(weatherStatusRegexp)).not.toBeInTheDocument();
    expect(screen.queryByText(temperatureRegexp)).not.toBeInTheDocument();

    await waitForElementToBeRemoved(screen.getAllByTitle('loading spinner'));

    expect(screen.queryByTitle('loading spinner')).not.toBeInTheDocument();
    expect(screen.queryByText(weatherStatusRegexp)).toBeInTheDocument();
    expect(screen.queryByText(temperatureRegexp)).toBeInTheDocument();
    expect(screen.queryByText(/не удалось загрузить данные о текущей погоде/i)).not.toBeInTheDocument();
  });
});
