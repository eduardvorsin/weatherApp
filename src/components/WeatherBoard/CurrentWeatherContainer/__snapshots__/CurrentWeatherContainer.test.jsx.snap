// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentWeatherContainer component tests snapshot with a spinner 1`] = `
<section
  class="current-weather__section current-weather-section test-class"
  data-testid="current-weather"
>
  <h2
    class="current-weather-section__title sr-only"
  >
    Текущая погода
  </h2>
  <div
    aria-label="loading spinner"
    class="spinner "
    style="--size: 102px;"
    title="loading spinner"
  >
    <div
      class="spinner__ring"
      style="--ring-size: 81.60000000000001px;"
    />
    <div
      class="spinner__ring"
      style="--ring-size: 81.60000000000001px;"
    />
    <div
      class="spinner__ring"
      style="--ring-size: 81.60000000000001px;"
    />
    <div
      class="spinner__ring"
      style="--ring-size: 81.60000000000001px;"
    />
  </div>
</section>
`;

exports[`CurrentWeatherContainer component tests snapshot with mock data 1`] = `
<section
  class="current-weather__section current-weather-section test-class"
  data-testid="current-weather"
>
  <h2
    class="current-weather-section__title sr-only"
  >
    Текущая погода
  </h2>
  <div
    class="current-weather "
    data-testid=""
  >
    <p
      class="current-weather__temperature"
    >
      -6.7°
    </p>
    <p
      class="current-weather__status"
    >
      ясно
    </p>
  </div>
</section>
`;

exports[`CurrentWeatherContainer component tests snapshot with mock error 1`] = `
<section
  class="current-weather__section current-weather-section test-class"
  data-testid="current-weather"
>
  <h2
    class="current-weather-section__title sr-only"
  >
    Текущая погода
  </h2>
  <div
    class="error-info  "
    data-testid=""
    role="alert"
  >
    <h3
      class="error-info__description"
    >
      Не удалось загрузить данные о текущей погоде
    </h3>
  </div>
</section>
`;

exports[`CurrentWeatherContainer component tests snapshot without anything 1`] = `
<section
  class="current-weather__section current-weather-section test-class"
  data-testid="current-weather"
>
  <h2
    class="current-weather-section__title sr-only"
  >
    Текущая погода
  </h2>
</section>
`;
